feat: Adds support for user defined subject token suppliers in AWSCredentials and IdentityPoolCredentials

aeitzman opened the pull request:
see go/programmable-auth-design.
Also adds quality of life improvements for building credentials directly via the builder methods, and a subject token type enum.

aeitzman commented on oauth2_http/java/com/google/auth/oauth2/AwsCredentials.java:
```diff
@@ -132,6 +156,9 @@ public GoogleCredentials createScoped(Collection<String> newScopes) {
   @Override
   String getCredentialSourceType() {
+    if (this.awsSecurityCredentialsSupplier != null) {
+      return "programmatic";
```
We can already tell that its AWS vs identity pool from the token type requested (AWSV4 vs JWT/Id_token/SAML2. will change to constants though

aeitzman commented on oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:
```diff
@@ -280,8 +294,6 @@ protected ExternalAccountCredentials(ExternalAccountCredentials.Builder builder)
             ? new ExternalAccountMetricsHandler(this)
             : builder.metricsHandler;
-
-    this.impersonatedCredentials = buildImpersonatedCredentials();
```
Sure, probably makes sense to just to keep it the same. I didn't make that change initially since the issue we have with the impersonated credentials is only possible in the builder constructor

lsirac commented:
We should also update the README here

aeitzman commented on oauth2_http/java/com/google/auth/oauth2/IdentityPoolCredentials.java:
```diff
@@ -53,6 +54,9 @@
+import javax.annotation.Nullable;
 
 /**
  * Url-sourced and file-sourced external account credentials.
```
added supplier info in javadoc for both identity pool and aws

aeitzman commented:
> We should also update the README here + add integration tests

added integration tests, will add README update in a separate PR

TimurSadykov commented on oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:
```diff
@@ -492,10 +503,10 @@ private static boolean isAwsCredential(Map<String, Object> credentialSource) {
-    if (impersonatedCredentialsOverride != null) {
-      return impersonatedCredentialsOverride.refreshAccessToken();
-    } else if (impersonatedCredentials != null) {
+    if (this.isImpersonationEnabled() && this.impersonatedCredentials == null) {
+      this.impersonatedCredentials = this.buildImpersonatedCredentials();
+    }
+    if (impersonatedCredentials != null) {
       return impersonatedCredentials.refreshAccessToken();
     }
```
This can look easier to understand and simpler if combined. Or does it mean that impersonatedCredentials can be provided with isImpersonationEnabled == false?

aeitzman replied:
Not sure exactly what you are implying with combined? you mean adding the this.impersonatedCredentials check  into the isImpersonationEnabled method? I can do that, would just rename the method shouldBuildImpersatedCredentials I guess. The idea of the check here is to only call the build method if we haven't already instantiated the impersonated credentials if that wasn't clear.

TimurSadykov commented on oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:
```diff
@@ -492,10 +503,10 @@ private static boolean isAwsCredential(Map<String, Object> credentialSource) {
-    if (impersonatedCredentialsOverride != null) {
-      return impersonatedCredentialsOverride.refreshAccessToken();
-    } else if (impersonatedCredentials != null) {
+    if (this.isImpersonationEnabled() && this.impersonatedCredentials == null) {
```
Why we can we have a impersonationUrl but impersonated credentials not ready? Can we build them whenever url is set?

aeitzman replied:
see my reply on your first comment, building the impersonated credential in the parent constructor has some issues.

TimurSadykov reviewed (commented):
Couple first comments, still looking

TimurSadykov commented:
> _ No description provided. _

@aeitzman please add description: reference to an internal doc. And a brief description, like your QoL changes. I know the context, but if other team members or Cloud Java folks will look - it is unclear. Also in the future it would be hard to uncover context.

aeitzman commented on oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:
```diff
@@ -104,11 +123,7 @@ abstract static class CredentialSource implements java.io.Serializable {
-  // Internal override for impersonated credentials. This is done to keep
-  // impersonatedCredentials final.
-  @Nullable private ImpersonatedCredentials impersonatedCredentialsOverride;
+  @Nullable protected ImpersonatedCredentials impersonatedCredentials;
```
The issue with having impersonated credentials final is that the ImpersonatedCredentials object is built by passing the current credential as the SourceCredential. If we have it final, the Impersonated Credential needs to get created in the parent constructor before the child constructor runs, which means the credential object that gets passed as the source credential is not fully instantiated. As a series of steps, the exact problem is:

1. AwsCredentials constructor is called via the builder
2. constructor super() is called, going to the External account credentials constructor
3. in the parent constructor, buildImpersonatedCredentials is called
4. the ImpersonatedCredentials object is built, but the source credential passed doesn't have any of the AwsCredentials specific builder variables set (AwsSecurityCredentialsSupplier for example) because the rest of the AwsCredentials constructor logic has not run yet, we are still in the parent constructor.

Since the source credential passed to the impersonated credential isn't actually correct here we have 1 of two options if we want any child class specific builder setters
1: Set the impersonatedCredentialOverride in the child constructor and just use that, ignoring the "final" impersonatedCredentials. This is what we do for executableCredentials, but the issue is that its kind of a weird pattern that we just have the unused and incorrect impersonatedCredentials object attached to the credential, and that it means we can't do any input validation on the child credential.
2: Make impersonatedCredentials not final, and just set it when we use it the first time when retrieving an access token. This lets us get rid of the override, and makes it so the impersonatedCredential is never built with the incorrect source credential.

aeitzman commented on oauth2_http/java/com/google/auth/oauth2/PluggableAuthCredentials.java:
```diff
@@ -192,7 +192,7 @@ public PluggableAuthCredentials createScoped(Collection<String> newScopes) {
   @Override
   String getCredentialSourceType() {
-    return "executable";
+    return PLUGGABLE_AUTH_METRICS_HEADER_VALUE;
```
that's what this method is used for, can rename the method if you feel strongly since its package private so wouldn't be a breaking change.

TimurSadykov replied:
Why not the PLUGGABLE_AUTH_CREDENTIAL_SOURCE_TYPE?

aeitzman commented on oauth2_http/java/com/google/auth/oauth2/ExternalAccountMetricsHandler.java:
```diff
@@ -43,7 +43,7 @@ class ExternalAccountMetricsHandler implements java.io.Serializable {
   private final boolean configLifetime;
   private final boolean saImpersonation;
-  private String credentialSourceType;
+  private ExternalAccountCredentials credentials;
```
This needed to change so I could keep the metrics handler final in External account credentials. Since the metric source could change due to things in child credential specific builder and we create the metrics handler in the parent class constructor.

aeitzman commented on oauth2_http/java/com/google/auth/oauth2/IdentityPoolCredentials.java:
```diff
@@ -53,22 +54,44 @@
-  private final IdentityPoolCredentialSource identityPoolCredentialSource;
+  @Nullable private final IdentityPoolCredentialSource identityPoolCredentialSource;
+
+  @Nullable private final Supplier<String> subjectTokenSupplier;
```
no, since its just a different way to retrieve the subject token. The only place we currently do any additional logic on the subject token after retrieving it is the AwsCredentials (because we take the AWS security credentials and turn it into a signed key). All other current use cases are covered by the logic added to identity pool credentials.

TimurSadykov commented on oauth2_http/java/com/google/auth/oauth2/AwsCredentials.java:
```diff
@@ -66,17 +71,50 @@ public class AwsCredentials extends ExternalAccountCredentials {
     super(builder);
-    this.awsCredentialSource = (AwsCredentialSource) builder.credentialSource;
+    // Check that one and only one of supplier or credential source are provided.
+    if (builder.awsSecurityCredentialsSupplier != null && builder.credentialSource != null) {
```
This and the rest of the if-else changes indicate, that you can split it. 

It seems to be cleaner to have a separate AwsProgrammaticCredential that extends this one. It can add its own source type, but hide setting the other one. And it can override whatever methods you change here. 

It is easier to understand, could this be an option?

TimurSadykov replied:
Similarly to IdentityPool.

lsirac replied:
Hey Timur, we appreciate your suggestions but we are not going to add new classes. The design has been approved and it is too late to make changes.

lsirac replied:
Discussed offline, we've agreed to go with @westarle's suggestions. Thanks guys.

aeitzman commented on oauth2_http/java/com/google/auth/oauth2/InternalAwsSecurityCredentialsProvider.java:
```diff
@@ -0,0 +1,262 @@
+    this.region = region;
+  }
+
+  public AwsSecurityCredentials getCredentials() throws IOException {
```
IDE was making me have this public since I guess all methods from the interface are required to be public, assuming this is why we had the mismatch for the credential sources as well. Will fix it by making the parent abstract instead of an interface

aeitzman replied:
Changed to abstract class, and made all package-private

aeitzman commented on oauth2_http/java/com/google/auth/oauth2/IdentityPoolSubjectTokenProvider.java:
```diff
@@ -0,0 +1,70 @@
+abstract class IdentityPoolSubjectTokenProvider implements Serializable {
+  abstract String getSubjectToken() throws IOException;
+
+  abstract boolean isUserSupplied();
```
Need it to figure out what metric header value to use (programmatic vs file/URL) but thinking about it more, it makes more sense for the provider to just return the metrics header value instead of a bool here

aeitzman replied:
Changed to metrics header, and added a separate bool in the actual credentials classes to keep track of whether we should be passing the supplier in the builder - I think that makes it more clear what's happening and obvious when making changes in the future.

aeitzman commented on oauth2_http/java/com/google/auth/oauth2/InternalAwsSecurityCredentialsProvider.java:
```diff
@@ -0,0 +1,262 @@
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+
+class InternalAwsSecurityCredentialsProvider implements AwsSecurityCredentialsProvider {
```
the naming of having the different stuff at the beginning makes more sense to me tbh, just seems inconsistent otherwise (I.e. we don't call it CredentialsAws). could put them in a folder I guess too which would make the file structure more readable

lsirac commented on oauth2_http/java/com/google/auth/oauth2/AwsSecurityCredentialsProvider.java:
```diff
@@ -35,13 +35,39 @@
+   *
+   * @return the Supplier used to retrieve the AWS security credentials.
+   */
+  abstract Supplier<AwsSecurityCredentials> getSupplier();
```
Why do we expose this? getCredentials() should be enough, right?

lsirac commented on oauth2_http/java/com/google/auth/oauth2/AwsCredentials.java:
```diff
@@ -103,13 +107,15 @@ public class AwsCredentials extends ExternalAccountCredentials {
       this.awsSecurityCredentialsProvider =
           new ProgrammaticAwsSecurityCredentialsProvider(
               builder.awsSecurityCredentialsSupplier, this.awsRegion);
+      this.builtWithSupplier = true;
```
Hmm, we shouldn't have to track this. Why not handle this in the builder so we can set the provider directly and not have to handle this logic here?

aeitzman replied:
Figured out a better way to handle in the builder, we can just check whether the builder has a credential source, and only set the supplier if it doesn't

lsirac commented on oauth2_http/java/com/google/auth/oauth2/IdentityPoolCredentials.java:
```diff
@@ -52,6 +52,9 @@ public class IdentityPoolCredentials extends ExternalAccountCredentials {
   private final IdentityPoolSubjectTokenProvider subjectTokenProvider;
+  // Boolean to tell if a Supplier was provided to build the credential. This is needed to tell
+  // whether the supplier should be passed when creating a new builder from this credential.
+  private final boolean builtWithSupplier;
```
We don't need track this -- instead of starting with builder.subjectTokenSupplier != null, move that to the else statement, and then rely on IdentityPoolCredentialSourceType.FILE and IdentityPoolCredentialSourceType.URL.

lsirac commented on oauth2_http/java/com/google/auth/oauth2/IdentityPoolCredentials.java:
```diff
@@ -136,7 +139,7 @@ public static class Builder extends ExternalAccountCredentials.Builder {
     Builder(IdentityPoolCredentials credentials) {
       super(credentials);
-      if (credentials.subjectTokenProvider.isUserSupplied()) {
+      if (credentials.builtWithSupplier) {
```
If the above change is made, we don't need this here (also, probably don't need to expose supplier at all).

lsirac commented on oauth2_http/java/com/google/auth/oauth2/IdentityPoolSubjectTokenProvider.java:
```diff
@@ -43,11 +43,32 @@
+   *
+   * @return a valid subject token.
+   * @throws IOException
+   */
```
Should we just call this getCredentials()?

westarle commented on oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:
```diff
@@ -488,10 +473,10 @@ private static boolean isAwsCredential(Map<String, Object> credentialSource) {
+    if (this.shouldBuildImpersonatedCredential()) {
+      this.impersonatedCredentials = this.buildImpersonatedCredentials();
+    }
+    if (impersonatedCredentials != null) {
```
Nit: consistency for `this.impersonatedCredentials` or just `impersonatedCredentials` here.

westarle commented on oauth2_http/java/com/google/auth/oauth2/IdentityPoolSubjectTokenProvider.java:
```diff
@@ -0,0 +1,91 @@
+ * Provider for retrieving subject tokens for {@Link IdentityPoolCredentials} to exchange for GCP
+ * access tokens.
+ */
+abstract class IdentityPoolSubjectTokenProvider implements Serializable {
```
nit: Interface?

westarle commented on oauth2_http/java/com/google/auth/oauth2/IdentityPoolSubjectTokenProvider.java:
```diff
@@ -0,0 +1,91 @@
+   *
+   * @return the supplier used for getSubjectToken.
+   */
+  abstract Supplier<String> getSupplier();
```
Maybe similar to @lsirac comment: we can probably remove getSupplier() and use getToken() directly (it would call the supplier in the programmatic case);

That said, it might be better if customers provided a `SubjectTokenProvider` directly.  The (internal) API allows us to cheat and deal with `IOException` in a way we don't allow via `Supplier<String>`;  I see both problems of us eating `IOException` and turning them into RTEs as well as us catching RTEs and turning them into not-retryable `IOException` -- even though we can anticipate that there will be network, IPC or File errors present based on all of our own implementations.

So...my suggestion is to ditch "Supplier" and offer "Provider" to our customers.  Now everyone has the same API `String getSubjectToken() throws IOException`; they can implement Retryable on their exceptions.  You might need to find another way to support the metrics header value if you're worried about clients overriding it, but overall the code should be simpler.

aeitzman replied:
Made this change - I wanted to keep the supplier so we could have a functional interface for users so they could use a lambda instead of creating a whole class, but I agree that its creating more problems than its worth. Refactored everything to be public. I named the objects IdentityPoolSubjectTokenSupplier and AwsSecurityCredentialSupplier instead of provider just because I call the similar function/object supplier in other PR's for other languages (still in flight, not merged) but don't feel super strongly about that either way.

westarle commented on oauth2_http/java/com/google/auth/oauth2/InternalAwsSecurityCredentialsProvider.java:
```diff
@@ -0,0 +1,282 @@
+  @VisibleForTesting
+  boolean shouldUseMetadataServer() {
+    return (!canRetrieveRegionFromEnvironment()
+        || !canRetrieveSecurityCredentialsFromEnvironment());
```
This can happen in the future, but:

now that you have an indirection to choose token providers, this logic can happen when installing a token provider and therefore break this class into one for environment and one for MDS.

aeitzman replied:
Unfortunately I think that would be a breaking change because the currently implementation checks the environment variables when get security credentials is called instead of in the constructor, so technically its possible for a user to create the credential and set environment variables after.

westarle commented on oauth2_http/java/com/google/auth/oauth2/IdentityPoolSubjectTokenProvider.java:
```diff
@@ -0,0 +1,91 @@
+   */
+  abstract Supplier<String> getSupplier();
+
+  protected static String parseToken(
```
Can this live with FileIDPoolCredential?

aeitzman replied:
Both file and URL need this, moved to file provider for now since it can't live in the interface and I don't think its worth creating a new class to hold 1 utility method.

westarle commented on oauth2_http/java/com/google/auth/oauth2/IdentityPoolCredentials.java:
```diff
@@ -88,76 +103,17 @@ public AccessToken refreshAccessToken() throws IOException {
-      return retrieveSubjectTokenFromCredentialFile();
-    }
-    return getSubjectTokenFromMetadataServer();
+    return this.subjectTokenProvider.getSubjectToken();
```
Here's an example of us cheating in our own implementation, offering `Retryable` `IOExceptions` for our own implementations but losing our customers RTEs into a generic `IOException`.

aeitzman replied:
I believe this is resolved by the refactor to have the supplier/provider be public to users but feel free to reopen if I'm not understanding correctly

westarle reviewed (commented):
This is looking pretty good now, can you please comment on the coverage of the change in unit and functional tests?

aeitzman commented:
> This is looking pretty good now, can you please comment on the coverage of the change in unit and functional tests?

Added unit tests for the builders using the new supplier objects, as well as tests for the new default values. Also added tests for the credentials built with the suppliers (happy path with/without service account impersonation and error paths). Did not add tests for the internal suppliers since those paths are covered by existing testing. Also added integration tests for AWS and IdentityPool subject token supplier built credentials.

lsirac commented on oauth2_http/java/com/google/auth/oauth2/IdentityPoolSubjectTokenSupplier.java:
```diff
@@ -31,50 +31,21 @@
-  Supplier<String> getSupplier() {
-    return this.subjectTokenSupplier;
-  }
+  String getSubjectToken() throws IOException;
```
Why not getCredentials() for a more consist API surface?

westarle replied:
my 2c: if it returns a credentials object yes, otherwise I prefer this descriptive name for the string.

lsirac reviewed (approved):
LGTM thanks Alex 💯

westarle reviewed (approved):
If @lsirac or another reviewer has covered the Java details for bugs etc. this looks good to me.
